packages/instrumentation-amqplib/src/amqplib.ts

```diff
@@ -316,64 +329,164 @@ export class AmqplibInstrumentation extends InstrumentationBase<typeof amqp> {
+                exchange,
+                self,
+                routingKey,
+                channel,
+                moduleVersion,
+                options
+            );
 
-            if (self._config.publishHook) {
+            if (self._config.publishConfirmHook) {
                 safeExecuteInTheMiddle(
-                    () => self._config.publishHook(span, { exchange, routingKey, content, options }),
+                    () => self._config.publishConfirmHook(span, { exchange, routingKey, content, options }),
                     (e) => {
                         if (e) {
-                            diag.error('amqplib instrumentation: publishHook error', e);
+                            diag.error('amqplib instrumentation: publishConfirmHook error', e);
                         }
                     },
                     true
                 );
             }
 
-            // calling original publish function is only storing the message in queue.
-            // it does not send it and waits for an ack, so the span duration is expected to be very short.
+            const patchedOnConfirm = function (err: any, ok: amqp.Replies.Empty) {
+                // should we wrap with context and end span after this callback or end span right away?
+                context.with(trace.setSpan(context.active(), span), () => {
+                    callback?.call(this, err, ok);
+                });
+
+                if (self._config.publishConfirmEndHook) {
+                    safeExecuteInTheMiddle(
+                        () => self._config.publishConfirmEndHook(span, { exchange, routingKey, content, options }, err),
+                        (e) => {
+                            if (e) {
+                                diag.error('amqplib instrumentation: publishConfirmEndHook error', e);
+                            }
+                        },
+                        true
+                    );
+                }
+
+                if (err) {
+                    span.setStatus({
+                        code: SpanStatusCode.ERROR,
+                        message: "message confirmation has been nack'ed",
+                    });
+                } else {
+                    span.setStatus({ code: SpanStatusCode.OK });
+                }
+                span.end();
+            };
+
+            // calling confirm channel publish function is storing the message in queue and registering the callback for broker confirm.
+            // span ends in the patched callback.
             const argumentsCopy = [...arguments];
             argumentsCopy[3] = modifiedOptions;
-            const originalRes = original.apply(this, argumentsCopy);
-            if (!originalRes) {
-                span.setStatus({
-                    code: SpanStatusCode.ERROR,
-                    message: 'write buffer is full, message should be published again after drain event',
-                });
+            argumentsCopy[4] = patchedOnConfirm;
```

Comment:
An alternative, probably a bit more efficient and idiomatic way of writing it would be to do something like:
```js
argumentsCopy[4] = context.bind(trace.setSpan(context.active(), span), patchedOnConfirm)
```

Up to you (just a suggestion for improvement)

Reply:
Fine for me, thanks